### PR TITLE
feat: openai streaming tool calls

### PR DESCRIPTION
**Pull request by axiomofjoy**
- adds tool call support for synchronous and asynchronous streaming
- adds corresponding unit tests

### CODE REVIEW

File: src/phoenix/trace/openai/instrumentor.py [hunk 1]
```diff
@@ -616,8 +626,12 @@ def _accumulate_messages(
     if not chunks:
         return []
     content_token_lists: DefaultDict[int, List[str]] = defaultdict(list)
-    function_argument_token_lists: DefaultDict[int, List[str]] = defaultdict(list)
-    function_names: Dict[int, str] = {}
+    function_calls: DefaultDict[int, StreamingFunctionCallData] = defaultdict(
+        lambda: StreamingFunctionCallData(argument_tokens=[])
+    )
+    tool_calls: DefaultDict[int, DefaultDict[int, StreamingFunctionCallData]] = defaultdict(
+        lambda: defaultdict(lambda: StreamingFunctionCallData(argument_tokens=[]))
```
Comment (RogerHYang) [inline src/phoenix/trace/openai/instrumentor.py:633]: nit. i think using dataclass can save you a few keystrokes
```suggestion
        lambda: defaultdict(StreamingFunctionCallData))
```

**Review by RogerHYang — APPROVED** [review]

File: src/phoenix/trace/openai/instrumentor.py [hunk 2]
```diff
@@ -626,21 +640,42 @@ def _accumulate_messages(
...
+        if content_tokens := content_token_lists[choice_index]:
+            message[MESSAGE_CONTENT] = "".join(content_tokens)
+        if function_call_ := function_calls.get(choice_index):
+            if (name := function_call_.get("name")) is not None:
+                message[MESSAGE_FUNCTION_CALL_NAME] = name
+            if argument_tokens := function_call_.get("argument_tokens"):
+                message[MESSAGE_FUNCTION_CALL_ARGUMENTS_JSON] = "".join(argument_tokens)
+        if tool_calls_ := tool_calls.get(choice_index):
+            message[MESSAGE_TOOL_CALLS] = [{} for _ in range(len(tool_calls_))]
+            for tool_index, tool_call_ in tool_calls_.items():
```
Comment (RogerHYang) [inline src/phoenix/trace/openai/instrumentor.py:672]: i think you may want to use enumerate instead to ward off any potential index out of bound errors
```suggestion
            for tool_index, tool_call_ in enumerate(tool_calls_.values()):
```

File: src/phoenix/trace/openai/instrumentor.py (hunk 1 above)
Comment (axiomofjoy) [reply src/phoenix/trace/openai/instrumentor.py]: Addressed in d9112733

File: src/phoenix/trace/openai/instrumentor.py (hunk 2 above)
Comment (axiomofjoy) [reply src/phoenix/trace/openai/instrumentor.py:674]: addressed in df18361a

File: src/phoenix/trace/openai/instrumentor.py [hunk 3]
```diff
@@ -626,21 +641,43 @@ def _accumulate_messages(
...
+                    function_calls[choice_index].name = function_name
                 if (function_argument_token := function_call.arguments) is not None:
-                    function_argument_token_lists[choice_index].append(function_argument_token)
+                    function_calls[choice_index].argument_tokens.append(function_argument_token)
             if role := choice.delta.role:
                 roles[choice_index] = role
-    messages: List[OpenInferenceMessage] = [{} for _ in range(num_choices)]
+            if choice.delta.tool_calls:
+                for tool_call in choice.delta.tool_calls:
+                    tool_index = tool_call.index
```
Comment (RogerHYang) [inline src/phoenix/trace/openai/instrumentor.py:651]: for extra safety?
```suggestion
                for tool_index, tool_call in enumerate(choice.delta.tool_calls):
```

**Pull request merged**